dvc/rwlock.py

```diff
@@ -0,0 +1,182 @@
+        )
+
+
+@contextmanager
+def _edit_rwlock(lock_dir):
+    path = os.path.join(lock_dir, "rwlock")
+    try:
+        with open(path, "r") as fobj:
+            lock = json.load(fobj)
+        lock = SCHEMA(convert_to_unicode(lock))
+    except FileNotFoundError:
+        lock = SCHEMA({})
+    except JSONDecodeError as exc:
+        raise RWLockFileCorruptedError(path, cause=exc)
+    except Invalid as exc:
+        raise RWLockFileFormatError(path, cause=exc)
+    lock = defaultdict(dict, lock)
+    lock["read"] = defaultdict(list, lock["read"])
+    lock["write"] = defaultdict(dict, lock["write"])
+    yield lock
+    with open(path, "w+") as fobj:
+        json.dump(lock, fobj)
+
+
+def _info_to_str(info):
+    return "  (PID {}): {}".format(info["pid"], info["cmd"])
+
+
+def _get_urls(lock, path_info):
+    ret = set(filter(path_info.overlaps, lock.keys()))
+    ret.add(path_info.url)
+    return ret
+
+
+def _check_no_writers(lock, info, path_infos):
+    for path_info in path_infos:
+        urls = _get_urls(lock["write"], path_info)
+        writers = [lock["write"].get(url) for url in urls]
+        writers = list(filter(bool, writers))
+        if not writers:
+            continue
+
+        assert len(writers) == 1
+        writer = writers[0]
+        if writer == info:
+            continue
+
+        raise LockError(
+            "'{}' is busy, it is being written to by:\n{}".format(
+                str(path_info), _info_to_str(writer)
+            )
+        )
+
+
+def _check_no_readers(lock, path_infos):
+    from itertools import chain
+
+    for path_info in path_infos:
+        urls = _get_urls(lock["read"], path_info)
+
+        readers = list(
+            chain.from_iterable([lock["read"].get(url, []) for url in urls])
+        )
+        if not readers:
+            continue
+
+        str_readers = "\n".join(_info_to_str(r) for r in readers)
+        raise LockError(
+            "'{}' is busy, it is being read by:\n{}".format(
+                str(path_info), str_readers
+            )
+        )
+
+
+def _acquire_read(lock, info, path_infos):
+    changes = []
+
+    for path_info in path_infos:
+        url = path_info.url
+
+        readers = lock["read"][url]
+        if info in readers:
+            continue
+
+        changes.append(url)
+        readers.append(info)
+
+    return changes
+
+
+def _acquire_write(lock, info, path_infos):
+    changes = []
+
+    for path_info in path_infos:
+        url = path_info.url
+
+        if lock["write"][url] == info:
+            continue
+
+        changes.append(url)
+        lock["write"][url] = info
+
+    return changes
+
+
+def _release_write(lock, info, changes):
+    for url in changes:
+        assert "write" in lock
+        assert url in lock["write"]
+        assert lock["write"][url] == info
+        del lock["write"][url]
+        if not lock["write"]:
+            del lock["write"]
```

Comment:
I would just leave this, why bother? Not that important though.

Reply:
Just so it looks nicer.